### PR TITLE
Add (cancelable) ToastAddEvent for use in ToastGui.add

### PR DESCRIPTION
**Pull request by Roundaround**
Gives modders the ability to hook into the ToastGui's toast queue. To prevent the toast from being added, cancel the event. To change the toast, modify it directly or replace it with event.setToast(IToast).

### CODE REVIEW

File: src/main/java/net/minecraftforge/client/event/ToastAddEvent.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
+package net.minecraftforge.client.event;
```
Comment (ChampionAsh5357) [inline src/main/java/net/minecraftforge/client/event/ToastAddEvent.java:20]: Missing license header.

File: src/main/java/net/minecraftforge/client/event/ToastAddEvent.java [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+import net.minecraft.client.gui.toasts.IToast;
+import net.minecraftforge.eventbus.api.Cancelable;
+import net.minecraftforge.eventbus.api.Event;
+
+@Cancelable
+public class ToastAddEvent extends Event {
```
Comment (ChampionAsh5357) [inline src/main/java/net/minecraftforge/client/event/ToastAddEvent.java]: Incorrect formatting, forge uses new line braces.

File: src/main/java/net/minecraftforge/client/event/ToastAddEvent.java [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+    private IToast toast;
+    public ToastAddEvent(IToast toast) {
```
Comment (ChampionAsh5357) [inline src/main/java/net/minecraftforge/client/event/ToastAddEvent.java]: New line braces.

File: src/main/java/net/minecraftforge/client/event/ToastAddEvent.java [hunk 4]
```diff
@@ -0,0 +1,21 @@
+package net.minecraftforge.client.event;
+
+import net.minecraft.client.gui.toasts.IToast;
+import net.minecraftforge.eventbus.api.Cancelable;
+import net.minecraftforge.eventbus.api.Event;
+
+@Cancelable
+public class ToastAddEvent extends Event {
+    private IToast toast;
```
Comment (ChampionAsh5357) [inline src/main/java/net/minecraftforge/client/event/ToastAddEvent.java]: Separate fields from constructor with empty line.

File: src/main/java/net/minecraftforge/client/event/ToastAddEvent.java [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+    public ToastAddEvent(IToast toast) {
+        setToast(toast);
+    }
+
+    public IToast getToast() {
```
Comment (ChampionAsh5357) [inline src/main/java/net/minecraftforge/client/event/ToastAddEvent.java]: New line braces.

File: src/main/java/net/minecraftforge/client/event/ToastAddEvent.java [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+        return toast;
+    }
+
+    public void setToast(IToast toast) {
```
Comment (ChampionAsh5357) [inline src/main/java/net/minecraftforge/client/event/ToastAddEvent.java]: New line braces.

**Review by ChampionAsh5357 — COMMENTED** [review]
Bunch of things that would probably need to be addressed.
- There is no test case. How do we know this works rather than just some assumption that it does.
- There is no documentation. How do I use this event? What does it do?
- This will break existing Minecraft behavior because it doesn't replace all cases of the toast. e.g. `Tutorial#addTimedToast`. This will break all tutorial steps if replaced.

File: patches/minecraft/net/minecraft/client/gui/toasts/ToastGui.java.patch [hunk 7]
```diff
@@ -1,13 +1,12 @@
...
+@@ -62,6 +_,9 @@
     }
  
     public void func_192988_a(IToast p_192988_1_) {
--      this.field_191792_h.add(p_192988_1_);
 +      net.minecraftforge.client.event.ToastAddEvent event = new net.minecraftforge.client.event.ToastAddEvent(p_192988_1_);
 +      if (net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event)) return;
-+      this.field_191792_h.add(event.getToast());
++      p_192988_1_ = event.getToast();
+       this.field_191792_h.add(p_192988_1_);
```
Comment (ChampionAsh5357) [inline patches/minecraft/net/minecraft/client/gui/toasts/ToastGui.java.patch]: Just added another line, revert to what you previously had.

**Comment by Roundaround** [comment]
I've formatted the files as requested, added JavaDoc comments to the event, and written a text/example mod demonstrating the event usage.

As for your third point about situations like `Tutorial#addTimedToast`, I've removed the ability to set the toast in the event (and therefore made it final), so that references to the passed toast remain valid.

**Comment by Roundaround** [comment]
Sorry, accidentally closed when I attempted to rename the branch. Apparently you can rename the target branch and PRs will update, but the same cannot be said for the source branch.

**Review by Roundaround — COMMENTED** [review]
Responded to feedback. Not sure if I have to mark it reviewed to bounce it back? I've never used GitHub's code review system before

File: src/test/java/net/minecraftforge/debug/client/ToastAddTest.java [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+        {
+            CustomToast customToast = (CustomToast) toast;
+            if (!customToast.getText().startsWith("Notice: "))
+            {
+                customToast.setText("Notice: " + customToast.getText());
+            }
+        }
+    }
+
+    @OnlyIn(Dist.CLIENT)
```
Comment (TheCurle) [inline src/test/java/net/minecraftforge/debug/client/ToastAddTest.java:60]: This is poor practice, and in actuality won't help prevent any issues.. this should be removed from the test mod.

**Review by TheCurle — CHANGES_REQUESTED** [review]
Seems good, and if you can fix the issue quickly we can merge this with the next batch of PRs.

**Comment by sciwhiz12** [comment]
Hello! @Roundaround, please leave a comment whether you wish to continue this PR and update it to the latest active version of 1.18 (or leave an explanation why this PR's changes are not applicable to 1.18). If we do not hear back from you after some time, this PR will be closed for inactivity.
If anyone else wishes to continue the work of this PR, please create a PR with a link to this PR for reference. 
Thank you for your contributions. 👋

**Pull request closed without merging**